File: tests/compat/asyncpg/smoke.py
import asyncio
import socket
import subprocess
import sys
import time
from datetime import date, datetime
from decimal import Decimal
from pathlib import Path
from uuid import UUID

import asyncpg


def free_local_port() -> int:
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.bind(("127.0.0.1", 0))
        return sock.getsockname()[1]


def wait_for_endpoint(port: int, timeout: float = 10.0) -> None:
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        try:
            with socket.create_connection(("127.0.0.1", port), timeout=0.25):
                return
        except OSError:
            time.sleep(0.025)
    raise RuntimeError(f"gpu-db-engine-server did not start on 127.0.0.1:{port}")


class Server:
    def __init__(self, repo_root: Path):
        self.repo_root = repo_root
        self.port = free_local_port()
        self.process = None

    def start(self) -> None:
        subprocess.run(
            ["cargo", "build", "-p", "gpu_db_server", "--bin", "gpu-db-engine-server"],
            cwd=self.repo_root,
            check=True,
        )
        binary = self.repo_root / "target" / "debug" / "gpu-db-engine-server"
        self.process = subprocess.Popen(
            [str(binary), "--listen", f"127.0.0.1:{self.port}"],
            cwd=self.repo_root,
            stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.PIPE,
            text=True,
        )
        try:
            wait_for_endpoint(self.port)
        except Exception:
            stderr = self.stop()
            raise RuntimeError(f"server failed to start\n{stderr}")

    def stop(self) -> str:
        if self.process is None:
            return ""
        if self.process.poll() is None:
            self.process.terminate()
            try:
                self.process.wait(timeout=5)
            except subprocess.TimeoutExpired:
                self.process.kill()
                self.process.wait(timeout=5)
        return self.process.stderr.read() if self.process.stderr is not None else ""


async def connect(port: int, application_name: str):
    return await asyncpg.connect(
        host="127.0.0.1",
        port=port,
        user="postgres",
        database="postgres",
        ssl=False,
        server_settings={"application_name": application_name},
    )


async def main(repo_root: Path) -> None:
    server = Server(repo_root)
    server.start()
    try:
        conn = await connect(server.port, "asyncpg_smoke")

        setup_status = await conn.execute(
            """
            CREATE TABLE asyncpg_people (id INT, name TEXT);
            INSERT INTO asyncpg_people (id, name)
            VALUES (1, 'Ada'), (2, 'Linus'), (3, 'Grace');
            """
        )
        assert setup_status == "INSERT 0 3"

        statement = await conn.prepare(
            "SELECT id, name FROM asyncpg_people WHERE id = $1"
        )
        rows = await statement.fetch(2)
        assert len(rows) == 1
        assert dict(rows[0]) == {"id": 2, "name": "Linus"}

        empty_rows = await statement.fetch(99)
        assert empty_rows == []

        await conn.execute(
            """
            CREATE TABLE asyncpg_all_types (
                row_id INT PRIMARY KEY, i2 SMALLINT, i4 INT, i8 BIGINT, amount NUMERIC(12,4),
                flag BOOL, note TEXT, day DATE, created_at TIMESTAMP, ident UUID
            )
            """
        )
        native_values = (
            1, -7, 42, -9, Decimal("12345.6700"), True, "Grüße",
            date(1999, 12, 31), datetime(2000, 1, 1, 0, 0, 1, 234567),
            UUID("550e8400-e29b-41d4-a716-446655440000"),
        )
        await conn.execute(
            "INSERT INTO asyncpg_all_types VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)",
            *native_values,
        )
        await conn.execute(
            "INSERT INTO asyncpg_all_types VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)",
            2, None, None, None, None, None, None, None, None, None,
        )
        all_types = await conn.fetchrow(
            "SELECT i2, i4, i8, amount, flag, note, day, created_at, ident "
            "FROM asyncpg_all_types WHERE row_id = 1"
        )
        assert tuple(all_types.values()) == native_values[1:]
        typed_nulls = await conn.fetchrow(
            "SELECT i2, i4, i8, amount, flag, note, day, created_at, ident "
            "FROM asyncpg_all_types WHERE row_id = 2"
        )
        assert tuple(typed_nulls.values()) == (None,) * 9

        await conn.execute("CREATE TABLE asyncpg_not_null_contract (id INT PRIMARY KEY)")
        await conn.execute("BEGIN")
        try:
            await conn.execute("INSERT INTO asyncpg_not_null_contract VALUES ($1)", None)
            raise AssertionError("PRIMARY KEY NULL unexpectedly succeeded")
        except asyncpg.PostgresError as error:
            assert error.sqlstate == "23502"
        try:
            await conn.fetchval("SELECT 1")
            raise AssertionError("constraint failure did not abort the explicit transaction")
        except asyncpg.PostgresError as error:
            assert error.sqlstate == "25P02"
        await conn.execute("ROLLBACK")
        await conn.execute("INSERT INTO asyncpg_not_null_contract VALUES ($1)", 1)

        try:
            await conn.execute("COPY asyncpg_people FROM STDIN WITH CSV HEADER DELIMITER ','")
            raise AssertionError("broader COPY CSV options unexpectedly succeeded")
        except asyncpg.PostgresError as error:
            assert error.sqlstate == "0A000"

        recovered = await conn.fetchrow(
            "SELECT name FROM asyncpg_people WHERE id = 3"
        )
        assert recovered["name"] == "Grace"

        await conn.close()

        pool = await asyncpg.create_pool(
            host="127.0.0.1",
            port=server.port,
            user="postgres",
            database="postgres",
            ssl=False,
            min_size=1,
            max_size=1,
            server_settings={"application_name": "asyncpg_pool_smoke"},
        )
        try:
            async with pool.acquire() as pooled_conn:
                pooled_setup = await pooled_conn.execute(
                    """
                    CREATE TABLE asyncpg_pool_check (id INT, name TEXT);
                    INSERT INTO asyncpg_pool_check (id, name) VALUES (1, 'pooled');
                    """
                )
                assert pooled_setup == "INSERT 0 1"
                pooled = await pooled_conn.fetchrow(
                    "SELECT id, name FROM asyncpg_pool_check WHERE id = $1", 1
                )
                assert dict(pooled) == {"id": 1, "name": "pooled"}

            reused = await pool.fetchrow(
                "SELECT name FROM asyncpg_pool_check WHERE id = $1", 1
            )
            assert reused["name"] == "pooled"
        finally:
            await pool.close()

        reconnected = await connect(server.port, "asyncpg_reconnect_smoke")
        try:
            reconnect_setup = await reconnected.execute(
                """
                CREATE TABLE asyncpg_reconnect_check (id INT, name TEXT);
                INSERT INTO asyncpg_reconnect_check (id, name) VALUES (1, 'reconnected');
                """
            )
            assert reconnect_setup == "INSERT 0 1"
            reconnect_simple = await reconnected.fetchrow(
                "SELECT id, name FROM asyncpg_reconnect_check WHERE id = $1", 1
            )
            assert dict(reconnect_simple) == {"id": 1, "name": "reconnected"}
        finally:
            await reconnected.close()
    finally:
        server.stop()


if __name__ == "__main__":
    asyncio.run(main(Path(sys.argv[1])))

File: tests/compat/node-postgres/smoke.mjs
import assert from 'node:assert/strict';
import { spawn, spawnSync } from 'node:child_process';
import net from 'node:net';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import pg from 'pg';

const { Client, Pool } = pg;

const scriptDir = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(scriptDir, '../../..');

function freeLocalPort() {
  return new Promise((resolvePort, reject) => {
    const server = net.createServer();
    server.once('error', reject);
    server.listen(0, '127.0.0.1', () => {
      const address = server.address();
      server.close(() => resolvePort(address.port));
    });
  });
}

function waitForEndpoint(port, timeoutMs = 10_000) {
  const deadline = Date.now() + timeoutMs;
  return new Promise((resolveReady, reject) => {
    const attempt = () => {
      const socket = net.createConnection({ host: '127.0.0.1', port });
      socket.once('connect', () => {
        socket.destroy();
        resolveReady();
      });
      socket.once('error', (error) => {
        socket.destroy();
        if (Date.now() >= deadline) {
          reject(new Error(`thunderbolt-db-server did not start on 127.0.0.1:${port}: ${error.message}`));
          return;
        }
        setTimeout(attempt, 25);
      });
    };
    attempt();
  });
}

async function startServer() {
  const build = spawnSync(
    'cargo',
    ['build', '-p', 'gpu_db_server', '--bin', 'thunderbolt-db-server'],
    { cwd: repoRoot, stdio: 'inherit' },
  );
  assert.equal(build.status, 0, 'cargo build for thunderbolt-db-server failed');

  const port = await freeLocalPort();
  const bin = resolve(repoRoot, 'target/debug/thunderbolt-db-server');
  const child = spawn(bin, ['--listen', `127.0.0.1:${port}`], {
    cwd: repoRoot,
    stdio: ['ignore', 'ignore', 'pipe'],
  });

  let stderr = '';
  child.stderr.setEncoding('utf8');
  child.stderr.on('data', (chunk) => {
    stderr += chunk;
  });

  try {
    await waitForEndpoint(port);
  } catch (error) {
    child.kill();
    throw new Error(`${error.message}\nserver stderr:\n${stderr}`);
  }

  return {
    port,
    async stop() {
      if (child.exitCode === null) {
        child.kill();
        await new Promise((resolveStopped) => child.once('exit', resolveStopped));
      }
    },
  };
}

function connectionConfig(port, applicationName) {
  return {
    host: '127.0.0.1',
    port,
    user: 'postgres',
    database: 'postgres',
    application_name: applicationName,
    ssl: false,
  };
}

function localIsoTimestamp(value) {
  const pad = (part, width = 2) => String(part).padStart(width, '0');
  return `${value.getFullYear()}-${pad(value.getMonth() + 1)}-${pad(value.getDate())}`
    + `T${pad(value.getHours())}:${pad(value.getMinutes())}:${pad(value.getSeconds())}.${pad(value.getMilliseconds(), 3)}Z`;
}

async function main() {
  const server = await startServer();
  try {
    const client = new Client(connectionConfig(server.port, 'node_postgres_smoke'));
    await client.connect();
    assert.ok(Number.isInteger(client.processID) && client.processID > 0);
    assert.ok(Number.isInteger(client.secretKey));

    const simple = await client.query('SELECT 1 AS one');
    assert.equal(simple.rows.length, 1);
    assert.equal(simple.rows[0].one, 1);

    await client.query(`
      CREATE TABLE node_people (id INT, name TEXT);
      INSERT INTO node_people (id, name)
      VALUES (1, 'Ada'), (2, 'Linus'), (3, 'Grace');
    `);

    const prepared = await client.query({
      name: 'node_people_by_id',
      text: 'SELECT id, name FROM node_people WHERE id = $1',
      values: [2],
    });
    assert.deepEqual(prepared.rows, [{ id: 2, name: 'Linus' }]);

    const empty = await client.query({
      name: 'node_people_by_id',
      text: 'SELECT id, name FROM node_people WHERE id = $1',
      values: [99],
    });
    assert.deepEqual(empty.rows, []);

    await client.query(`
      CREATE TABLE node_all_types (
        row_id INT PRIMARY KEY, i2 SMALLINT, i4 INT, i8 BIGINT, amount NUMERIC(12,4),
        flag BOOL, note TEXT, day DATE, created_at TIMESTAMP, ident UUID
      )
    `);
    const day = '1999-12-31';
    // node-postgres has no date-only/timestamp-without-time-zone native value; its normal
    // PostgreSQL representation for those columns is their ISO text form.
    const createdAt = '2000-01-01 00:00:01.234567';
    const ident = '550e8400-e29b-41d4-a716-446655440000';
    await client.query({
      name: 'node_all_types_insert',
      text: 'INSERT INTO node_all_types VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)',
      values: [1, -7, 42, '-9', '12345.6700', true, 'Grüße', day, createdAt, ident],
    });
    await client.query({
      name: 'node_all_types_insert',
      text: 'INSERT INTO node_all_types VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)',
      values: [2, null, null, null, null, null, null, null, null, null],
    });
    const allTypes = await client.query(
      'SELECT i2, i4, i8, amount, flag, note, day, created_at, ident FROM node_all_types WHERE row_id = $1',
      [1],
    );
    assert.deepEqual(
      {
        i2: allTypes.rows[0].i2,
        i4: allTypes.rows[0].i4,
        i8: allTypes.rows[0].i8,
        amount: allTypes.rows[0].amount,
        flag: allTypes.rows[0].flag,
        note: allTypes.rows[0].note,
        day: localIsoTimestamp(allTypes.rows[0].day).slice(0, 10),
        createdAt: localIsoTimestamp(allTypes.rows[0].created_at),
        ident: allTypes.rows[0].ident,
      },
      {
        i2: -7,
        i4: 42,
        i8: '-9',
        amount: '12345.6700',
        flag: true,
        note: 'Grüße',
        day,
        createdAt: '2000-01-01T00:00:01.234Z',
        ident,
      },
    );
    const typedNulls = await client.query(
      'SELECT i2, i4, i8, amount, flag, note, day, created_at, ident FROM node_all_types WHERE row_id = $1',
      [2],
    );
    assert.ok(Object.values(typedNulls.rows[0]).every((value) => value === null));

    await client.query(`
      CREATE TABLE node_check_contract (
        id INT PRIMARY KEY, amount NUMERIC(4,2),
        CONSTRAINT node_check_positive CHECK (amount > 0.00)
      )
    `);
    await client.query('BEGIN');
    await assert.rejects(
      () => client.query('INSERT INTO node_check_contract VALUES ($1, $2)', [1, '-1.00']),
      (error) => error.code === '23514',
      'CHECK violation must preserve SQLSTATE 23514',
    );
    await assert.rejects(
      () => client.query('SELECT 1'),
      (error) => error.code === '25P02',
      'CHECK failure must abort the explicit transaction',
    );
    await client.query('ROLLBACK');
    await client.query('INSERT INTO node_check_contract VALUES ($1, $2)', [1, '1.00']);

    await assert.rejects(
      () => client.query("COPY node_people FROM STDIN WITH CSV HEADER DELIMITER ','"),
      (error) => error.code === '0A000',
      'broader COPY CSV options should stay explicitly unsupported',
    );

    const recovered = await client.query('SELECT name FROM node_people WHERE id = 3');
    assert.deepEqual(recovered.rows, [{ name: 'Grace' }]);

    await client.end();

    const pool = new Pool({
      ...connectionConfig(server.port, 'node_postgres_pool_smoke'),
      max: 1,
    });
    const pooled = await pool.query('SELECT 1 AS one');
    assert.equal(pooled.rows[0].one, 1);
    await pool.end();

    const reconnected = new Client(connectionConfig(server.port, 'node_postgres_reconnect_smoke'));
    await reconnected.connect();
    const reconnectSimple = await reconnected.query('SELECT 1 AS one');
    assert.equal(reconnectSimple.rows[0].one, 1);
    await reconnected.end();
  } finally {
    await server.stop();
  }
}

await main();
